tests/test_secondary.py

```diff
@@ -0,0 +1,721 @@
+  secondary_ecu_key = uptane.common.canonical_key_from_pub_and_pri(
+      key_pub, key_pri)
+
+  # Load the public timeserver key.
+  key_timeserver_pub = demo.import_public_key('timeserver')
+  key_timeserver_pri = demo.import_private_key('timeserver')
+
+  # Load the public director key.
+  key_directortargets_pub = demo.import_public_key('director')
+
+  # Generate a trusted initial time for the Secondary.
+  clock = tuf.formats.unix_timestamp_to_datetime(int(time.time()))
+  clock = clock.isoformat() + 'Z'
+  tuf.formats.ISO8601_DATETIME_SCHEMA.check_match(clock)
+
+  # Set up client directories for the two Secondaries, containing the
+  # initial root.json and root.der (both, for good measure) metadata files
+  # so that the clients can validate further metadata they obtain.
+  # NOTE that running multiple clients in the same Python process does not
+  # work normally in the reference implementation, as the value of
+  # tuf.conf.repository_directories is client-specific, and it is set during
+  # uptane.common.create_directory_structure_for_client, and used when a
+  # client is created (initialization of a Secondary in our case)
+  # We're going to cheat in this test module for the purpose of testing
+  # and update tuf.conf.repository_directories before each Secondary is
+  # created,  to refer to the client we're creating.
+  for client_dir in [TEMP_CLIENT_DIR_1, TEMP_CLIENT_DIR_2, TEMP_CLIENT_DIR_3]:
+    uptane.common.create_directory_structure_for_client(
+        client_dir,
+        TEST_PINNING_FNAME,
+        {'imagerepo': TEST_IMAGE_REPO_ROOT_FNAME,
+        'director': TEST_DIRECTOR_ROOT_FNAME})
+
+
+
+
+
+def tearDownModule():
+  """This is run once for the full module, after all tests."""
+  destroy_temp_dir()
+
+
+
+
+
+class TestSecondary(unittest.TestCase):
+  """
+  "unittest"-style test class for the Secondary module in the reference
+  implementation
+
+  Note that these tests are NOT entirely independent of each other.
+  Several of them build on the results of previous tests. This is an unusual
+  pattern but saves code and works at least for now.
+  """
+
+  def test_01_init(self):
+    """
+    Tests uptane.clients.secondary.Secondary::__init__()
+    Note that this doesn't test the root files provided to the constructor, as
+    those aren't used at all in the initialization; those will be tested by
+    attempting an update in the test for process_metadata below.
+    """
+    # Initialization should be done here, in the test of __init__, not earlier,
+    # and we will use the clients set up here in further testing (rather than
+    # creating a new client/clients in every single test).
+    # Consequently, these variables are stuck being set here, and so are stuck
+    # as globals.
+    global secondary_instance_1
+    global secondary_instance_2
+    global secondary_instance_3
+
+    # TODO: Test with invalid pinning file
+    # TODO: Test with pinning file lacking a Director repo.
```

Comment:
It's slightly different: Director metadata being missing (tested already) vs pinning file failing to list a Director repository.